From a9c6a15de0dc56d1f104174cbce56c34aebbc972 Mon Sep 17 00:00:00 2001
From: Anton Mikanovich <amikan@ilbers.de>
Date: Wed, 7 Jul 2021 19:38:50 +0300
Subject: [PATCH 2/7] image: Add reference counter

Image generation can be also made in parallel, so we need to control
the usage of image_do_mounts/image_undo_mounts.

Signed-off-by: Anton Mikanovich <amikan@ilbers.de>
---
 meta/classes/image.bbclass | 39 ++++++++++++++++++++++++++++++++++----
 1 file changed, 35 insertions(+), 4 deletions(-)

diff --git a/meta/classes/image.bbclass b/meta/classes/image.bbclass
index 12d1616..b2a828c 100644
--- a/meta/classes/image.bbclass
+++ b/meta/classes/image.bbclass
@@ -52,23 +52,54 @@ DEPENDS += "${IMAGE_INSTALL}"
 ISAR_RELEASE_CMD_DEFAULT = "git -C ${LAYERDIR_core} describe --tags --dirty --match 'v[0-9].[0-9]*'"
 ISAR_RELEASE_CMD ?= "${ISAR_RELEASE_CMD_DEFAULT}"
 
+IMG_MOUNT_CNT = "${BUILDCHROOT_DIR}_image.mount"
+
 image_do_mounts() {
-    sudo flock ${MOUNT_LOCKFILE} -c ' \
+    sudo -s <<'EOSUDO'
+        ( flock 9
+        set -e
+
+        count="1"
+        if [ -f "${IMG_MOUNT_CNT}" ]; then
+            count=$(($(cat "${IMG_MOUNT_CNT}") + 1))
+        fi
+        echo $count > "${IMG_MOUNT_CNT}"
+        if [ $count -gt 1 ]; then
+            exit 0
+        fi
+
         mkdir -p "${BUILDROOT_DEPLOY}" "${BUILDROOT_ROOTFS}" "${BUILDROOT_WORK}"
         mount --bind "${DEPLOY_DIR_IMAGE}" "${BUILDROOT_DEPLOY}"
         mount --bind "${IMAGE_ROOTFS}" "${BUILDROOT_ROOTFS}"
         mount --bind "${WORKDIR}" "${BUILDROOT_WORK}"
-    '
+
+        ) 9>'${MOUNT_LOCKFILE}'
+EOSUDO
     buildchroot_do_mounts
 }
 
 image_undo_mounts() {
     buildchroot_undo_mounts
-    sudo flock ${MOUNT_LOCKFILE} -c ' \
+    sudo -s <<'EOSUDO'
+        ( flock 9
+        set -e
+
+        count="0"
+        if [ -f "${IMG_MOUNT_CNT}" ]; then
+            count=$(($(cat "${IMG_MOUNT_CNT}") - 1))
+            echo $count > "${IMG_MOUNT_CNT}"
+        fi
+        if [ $count -gt 0 ]; then
+            exit 0
+        fi
+        rm -f "${IMG_MOUNT_CNT}"
+
         umount "${BUILDROOT_DEPLOY}"
         umount "${BUILDROOT_ROOTFS}"
         umount "${BUILDROOT_WORK}"
-    '
+
+        ) 9>'${MOUNT_LOCKFILE}'
+EOSUDO
 }
 
 ROOTFSDIR = "${IMAGE_ROOTFS}"
-- 
2.20.1

